### ISSUE TITLE
ssh hangs during login

### ISSUE THREAD

**Issue by lumean** [net-ssh/net-ssh#356]
Hi,
Beginning with version 3.1.1 i have the issue that net-ssh gets stuck during login to a cisco device:

```
D, [2016-04-20T17:28:22.422385 #7518] DEBUG -- net.ssh.transport.session[bfcb9c]: establishing connection to 10.0.0.1:22
D, [2016-04-20T17:28:22.424337 #7518] DEBUG -- net.ssh.transport.session[bfcb9c]: connection established
I, [2016-04-20T17:28:22.424408 #7518]  INFO -- net.ssh.transport.server_version[bfc304]: negotiating protocol version
D, [2016-04-20T17:28:22.424431 #7518] DEBUG -- net.ssh.transport.server_version[bfc304]: local is `SSH-2.0-Ruby/Net::SSH_4.0.0.alpha3 x86_64-linux'
D, [2016-04-20T17:28:22.426128 #7518] DEBUG -- net.ssh.transport.server_version[bfc304]: remote is `SSH-2.0-Cisco-1.25'
```

attached is the logfile of 3.0.1 where the behavior is better but still the channel request fails. I got the same behavior also with 3.0.2
[ssh-3.0.1.txt](https://github.com/net-ssh/net-ssh/files/228262/ssh-3.0.1.txt)

From a normal linux ssh client I can connect to that cisco device without any issues.
Let me know if you need more debugging output.

**Comment by mfazekas** [comment]
So this sounds like 4.0.0.alpha3? 
is this the same with 3.1.1?
Do we have logs on device?
Also what is the backtrace it stuck?

It looks version negotiation was done, but server doesn't sends KEXINIT. Which is weird.

**Comment by lumean** [comment]
4.0.0.alpha3 behaves the same as 3.1.1.

attached you can find a comparison of 4.0.0.alpha3 and 3.0.1 where just the channel request fails - which is also weird because the server sais the tty / shell were successfully allocated (at least as far as my understanding goes)
[ssh_server_client_ruby_3.0.1.txt](https://github.com/net-ssh/net-ssh/files/229676/ssh_server_client_ruby_3.0.1.txt)
[ssh_server_client_ruby_400alpha3.txt](https://github.com/net-ssh/net-ssh/files/229677/ssh_server_client_ruby_400alpha3.txt)

**Comment by lumean** [comment]
In addition the behavior described above just applies to linux (centos7 in my case with ruby 2.2.3p173 (2015-08-18 revision 51636) [x86_64-linux]) 
On my windows 10 system with same ruby version ( ruby 2.2.3p173 (2015-08-18 revision 51636) [i386-mingw32])  i get a different behavior. See the attached debug outputs.
[ssh_server_client_ruby_3.1.1_centos7.txt](https://github.com/net-ssh/net-ssh/files/230098/ssh_server_client_ruby_3.1.1_centos7.txt)
[ssh_server_client_ruby_3.1.1_windows10.txt](https://github.com/net-ssh/net-ssh/files/230099/ssh_server_client_ruby_3.1.1_windows10.txt)

**Comment by mfazekas** [comment]
Thanks much. That is really weird. It looks like that cisco doesn't like when we send ssh version before server sends his. This was case of ssh1, but in ssh2 it should not matter.
Can you also try ```ssh -v -2 user@host``` that forces v2 of ssh.

```
Apr 21 15:52:42.762 CEST: SSH1: starting SSH control process
Apr 21 15:52:42.762 CEST: SSH1: sent protocol version id SSH-2.0-Cisco-1.25
Apr 21 15:52:42.773 CEST: SSH1: protocol version id is - SSH-2.0-Ruby/Net::SSH_3.1.1 i386-mingw32
```

```
Apr 21 15:56:36.102 CEST: SSH1: starting SSH control process
Apr 21 15:56:36.102 CEST: SSH1: sent protocol version id SSH-2.0-Cisco-1.25

Apr 21 15:58:36.103 CEST: SSH1: receive failure - status 0x12
Apr 21 15:58:36.202 CEST: SSH1: Session disconnected - error 0x12
```

**Comment by lumean** [comment]
Here is the debug output from the native linux client with the cisco server. that client can login normally and (I activated more debugging flags on the server
[ssh_server_client_native_centos7.txt](https://github.com/net-ssh/net-ssh/files/231732/ssh_server_client_native_centos7.txt)
, so debugs are a bit more verbose)

**Comment by mfazekas** [comment]
was this with ```ssh -v -2 user@host```? Does that connects 10 out of 10 times fine?

**Comment by lumean** [comment]
yes, with a linux client or also puTTY / SecureCRT I've never had any issues. I just tried and connected 10x in a row from linux to that cisco device, without problems.

**Comment by mfazekas** [comment]
But were you using the ```-2``` option to force ssh protocol version 2?

**Comment by lumean** [comment]
yes, I was using exactly this command "ssh -v -2 user@host"  ten times in a row

**Comment by lumean** [comment]
Could you point me to the source where we send ssh version before server sends his? Maybe I could try to change the behavior in my local environment and see whether this solves the issue.

**Comment by mfazekas** [comment]
This was the change 
https://github.com/net-ssh/net-ssh/pull/317/files
I'm thinking of adding an env var to enable ssh1 style handshake. I'll also doublecheck but in ssh2 we should be able to send before server sent.

**Comment by lumean** [comment]
I can confirm that it works when I revert that PR. But still it is weird that windows and linux behave differently. Maybe that has something to do with timing and the socket implementation.

**Comment by lumean** [comment]
Another small updated regarding the channel_request failed, this was a problem of my own code... I tried to open two channels which is not supported by this server. No follow up is needed for that.

**Comment by mfazekas** [comment]
Thanks much for the debugging i really appreciate it!

With -v2 flag 'ssh' also should send version first: https://github.com/openssh/openssh-portable/blob/master/sshconnect.c#L563-L571

Windows i assume just got lucky receiving first. But it's strange that Cisco changes it's logic based on when the client version was received.

What i can do is to send kexinit before server sends his:
https://github.com/net-ssh/net-ssh/pull/359
something like that should do.

**Comment by lumean** [comment]
glad that I can help. I quickly verified and  #359 also works for me.

**Comment by mfazekas** [comment]
should be fixed in 4.0.0.alpha3 and later.
https://github.com/net-ssh/net-ssh/commit/9115c32e93ee574f4f3bfd68252ffc93e485249b
